### DISCUSSION TITLE
Creating JSON files for jsPSych experiments with R

### DISCUSSION THREAD

**Discussion by JAQuent** [jspsych/jsPsych#1526] [Show and tell]
Hey,
I thought some might like this pipeline. What I like do for my jsPsych experiments, is to code all the trial information in R and hardcode that information into a .js file that I than load before the experiment. 

For instance, I can take a data.frame in R like this one:
```
# My data.frame with all experimental variables
df <- data.frame(a = 1:2, b = 3:4)

# Creating a .js file from this
json_strings_from_df(df, 'example.js') 
```
This particular function can be found [here](https://github.com/JAQuent/assortedRFunctions/blob/master/R/json_strings_from_df.R). This example .js file looks like this:

```
var a = "[1,2]";

a = JSON.parse(a);

var b = "[3,4]";

b = JSON.parse(b);
```

When I now run this 
`<script src="scene_repetition2_studyTask_trials.js"></script>`
in the `<head></head>` part, then I can use variables a and b in my experiment. If I want I can use them as timeline variables. For that,  I just create an array and use a simple for loop to assign the values to that array.

```
var trials = new Array();
var nTrial = a.length();
for (i = 0; i < nTrial; i++){
trials[i] = {a[i],
	         data: {b: b[i]};

};
```

then this arrary can be used like this

```
  var blank = {
    type: 'html-keyboard-response',
    stimulus: '',
    choices: jsPsych.NO_KEYS,
    data: jsPsych.timelineVariable('data'),
    trial_duration: jsPsych.timelineVariable('a'),
  };

```

and then it has to be added to the timeline like this

```
  var experiment = {
    timeline: [blank],
    timeline_variables: trials,
  };
  timeline.push(experiment);
```

If I have multiple counter balancing conditions, I use lists like this:
```
a1 <- c(1, 2)
a2 <- c(4, 5)
a  <- list(a1, a2)

create_json_variable_str('a', a)
```
This creates a string like this

`"var a = \"[[1,2],[4,5]]\";"`  and I just have to write this one to .js file and add a = JSON.parse(a); and I can access the correct a for my particular condition by choosing a[0] vs. a[1]. 

The function can be found [here](https://github.com/JAQuent/assortedRFunctions/blob/master/R/create_json_variable_str.R). 

Note sure if that's the most elegant way to handle this kind of stuff but since I am more proficient in R prefer this and maybe it's helpful for someone else.

**Comment by becky-gilbert** [comment]
This is very cool. Thanks for sharing!

**Comment by cvanheer** [comment]
Just wanted to say thanks cos I simulate my trials in R as well and was tossing up between doing csv -> json -> js but I realise I can just do the json bit in R as well so thank you :) @JAQuent!
